libsql-sqlite3/src/wal.c

```diff
@@ -1729,6 +1761,51 @@ static int walIteratorInit(Wal *pWal, u32 nBackfill, WalIterator **pp){
   return rc;
 }
 
+static int walIteratorRevInit(Wal *pWal, u32 nBackfill, struct WalIteratorRev *p, u32 mxSafeFrame, int getMissingFrame){
+    WalIterator *pIter;
+    u32 *frames;
+    u32 iFrame, iPageno;
+    int rc;
+
+    frames = (u32*)sqlite3MallocZero((pWal->hdr.mxFrame + 1) * sizeof(u32));
+    if (!frames) return SQLITE_NOMEM_BKPT;
+    rc = walIteratorInit(pWal, nBackfill, &pIter);
+    if (rc || !pIter) {
+        sqlite3_free(frames);
+        return rc;
+    }
+
+    while (walIteratorNext(pIter, &iPageno, &iFrame) == 0) {
+        /*
+         * If we get a page with a frame_no greater than mxSafeFrame, and getMissingFrame is true,
+         * then we replace it with the latest page with frame_no <= mxSafeFrame.
+         */
+        if (iFrame > mxSafeFrame && getMissingFrame) {
+           rc = walFindFrame(pWal, iPageno, mxSafeFrame, &iFrame);
+           if( rc!=SQLITE_OK ) break;
+           if (iFrame == 0) {
+               continue;
+           }
+        }
+        frames[iFrame] = iPageno;
+    }
+    walIteratorFree(pIter);
+    if (rc != 0) {
+        sqlite3_free(frames);
+        return rc;
+    }
+
+    p->current = pWal->hdr.mxFrame;
+    p->frames = frames;
+
+    return SQLITE_OK;
+}
+
+static void walIteratorRevFree(struct WalIteratorRev *p) {
+    p->current = 0;
+    sqlite3_free(p->frames);
```

Comment:
again, is it feasible to use `sqlite3DbFree` from the call site? It should be preferred whenever available